### PR TITLE
add logger and logs

### PR DESCRIPTION
**Pull request by CohenIdo**
add logger to APIs

### CODE REVIEW

File: x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.test.ts [hunk 1]
```diff
@@ -6,18 +6,21 @@
...
 import { elasticsearchClientMock } from 'src/core/server/elasticsearch/client/mocks';
+import { loggingSystemMock } from 'src/core/server/mocks';
 import { createFindingsIndexTemplate } from './create_index_template';
 const mockEsClient = elasticsearchClientMock.createClusterClient().asScoped().asInternalUser;
 afterEach(() => {
   mockEsClient.indices.putIndexTemplate.mockClear();
   mockEsClient.indices.existsIndexTemplate.mockClear();
 });
 describe('create index template for findings', () => {
+  const logger: ReturnType<typeof loggingSystemMock.createLogger> =
```
Comment (kfirpeled) [inline x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.test.ts:17]: looking on other examples, the initialization of the mock is happening in the `beforeeach` step
I think it is a good practice
wdyt?

File: x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts [hunk 2]
```diff
@@ -43,12 +44,14 @@ const createIndexTemplate = async (
     });
     return response.body.acknowledged;
   } catch (err) {
+    logger.error(`Invalid usage in putIndexTemplate for index ${indexName}`);
```
Comment (kfirpeled) [inline x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts:47]: please log the `err` as well

better text for the log is
`Failed to create index pattern ${indexName}, err`

Comment (kfirpeled) [reply x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts:47]: I saw now this text is already used
So maybe `putIndexTemplate failed`

Comment (kfirpeled) [reply x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts:47]: btw, I see todo:fix types.
what about it?

File: x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts [hunk 3]
```diff
@@ -57,11 +60,11 @@ export const createFindingsIndexTemplate = async (
...
       CSP_FINDINGS_INDEX_NAME,
       CSP_KUBEBEAT_INDEX_NAME,
-      // TODO: check why this cast is required
-      findingsIndexMapping
+      findingsIndexMapping,
+      logger
     );
   } catch (err) {
-    // TODO: add logger
+    logger.error(`Failed to create index template ${CSP_FINDINGS_INDEX_NAME}`);
```
Comment (kfirpeled) [inline x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts:67]: log the `err` as well

File: x-pack/plugins/cloud_security_posture/server/routes/stats/stats.ts [hunk 4]
```diff
@@ -48,6 +49,9 @@ const calculatePostureScore = (total: number, passed: number, failed: number): S
 const getLatestCycleId = async (esClient: ElasticsearchClient) => {
   const latestFinding = await esClient.search<LastCycle>(getLatestFindingQuery());
   const lastCycle = latestFinding.body.hits.hits[0];
+  if (lastCycle?._source?.run_id === undefined) {
```
Comment (kfirpeled) [inline x-pack/plugins/cloud_security_posture/server/routes/stats/stats.ts:52]: how does it relates to this pr?

File: x-pack/plugins/cloud_security_posture/server/routes/stats/stats.ts [hunk 5]
```diff
@@ -56,7 +60,10 @@ export const getBenchmarks = async (esClient: ElasticsearchClient) => {
   const benchmarksBuckets = queryResult.body.aggregations?.benchmarks as AggregationsTermsAggregate<
     DictionaryResponseBase<string, string>
   >;
-  return benchmarksBuckets.buckets.map((e) => e.key);
+  if (benchmarksBuckets?.buckets === undefined) {
```
Comment (kfirpeled) [inline x-pack/plugins/cloud_security_posture/server/routes/stats/stats.ts:63]: related to this pr?

**Review by kfirpeled — CHANGES_REQUESTED** [review]

File: x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts (hunk 2 above)
Comment (CohenIdo) [reply x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts:47]: A variable that defines under `beforeEach` is not available for the test. 
`beforeEach` contains mainly methods that we want o do before each test, in this case, the logger is used inside the test itself.

File: x-pack/plugins/cloud_security_posture/server/routes/stats/stats.ts (hunk 5 above)
Comment (CohenIdo) [reply x-pack/plugins/cloud_security_posture/server/routes/stats/stats.ts:63]: Same one^

File: x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.test.ts (hunk 1 above)
Comment (CohenIdo) [reply x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.test.ts:17]: A variable that defines under beforeEach is not available for the test.
beforeEach contains mainly methods that we want o do before each test, in this case, the logger is used inside the test itself.

File: x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts (hunk 2 above)
Comment (CohenIdo) [reply x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts:47]: Done:)

File: x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.test.ts (hunk 1 above)
Comment (CohenIdo) [reply x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.test.ts:17]: If I move the logger initialization to `beforeEach` I would also move the usage of the logger to `beforeEach` and I am afraid is not our use case.  
In `task_manager` plugin they are doing the same -> `x-pack/plugins/task_manager/server/monitoring/capacity_estimation.test.ts`
Another alternative that I saw is to define them inside the test itself.

File: x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts [hunk 6]
```diff
@@ -64,7 +64,7 @@ export const createFindingsIndexTemplate = async (
       logger
     );
   } catch (err) {
-    logger.error(`Failed to create index template ${CSP_FINDINGS_INDEX_NAME}`);
+    logger.error(`Failed to create index template ${CSP_FINDINGS_INDEX_NAME}, ${err}`);
```
Comment (kfirpeled) [inline x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts:67]: from a look in the kibana code it seems the common usage is
```js
logger.error('error message here')
logger.error(err)
```

Comment (kfirpeled) [reply x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts:67]: and worth adding debug log message whenever you are about to create something

File: x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.test.ts (hunk 1 above)
Comment (kfirpeled) [reply x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.test.ts:17]: attaching example

```js
describe('configureClient', () => {
  let logger: ReturnType<typeof loggingSystemMock.createLogger>;
  let config: ElasticsearchClientConfig;

  beforeEach(() => {
    logger = loggingSystemMock.createLogger();
    config = createFakeConfig();
    parseClientOptionsMock.mockReturnValue({});
    ClientMock.mockImplementation(() => createFakeClient());
  });
```

the logger is available this way for the tests

File: x-pack/plugins/cloud_security_posture/server/plugin.ts [hunk 7]
```diff
@@ -42,16 +42,18 @@ export class CspPlugin
   ): CspServerPluginSetup {
     this.logger.debug('csp: Setup');
     const router = core.http.createRouter();
-
-    defineRoutes(router);
+    // Register server side APIs
```
Comment (kfirpeled) [inline x-pack/plugins/cloud_security_posture/server/plugin.ts:45]: add line of space before a comment

File: x-pack/plugins/cloud_security_posture/server/routes/index.ts [hunk 8]
```diff
@@ -5,11 +5,11 @@
...
 
+import type { IRouter, Logger } from '../../../../../src/core/server';
 import { defineGetStatsRoute } from './stats/stats';
 import { defineFindingsIndexRoute as defineGetFindingsIndexRoute } from './findings/findings';
-import type { IRouter } from '../../../../../src/core/server';
 
-export function defineRoutes(router: IRouter) {
-  defineGetStatsRoute(router);
-  defineGetFindingsIndexRoute(router);
+export function defineRoutes(router: IRouter, logger: Logger) {
```
Comment (kfirpeled) [inline x-pack/plugins/cloud_security_posture/server/routes/index.ts:12]: example i saw today:

```js
export interface RouteDefinitionParams {
  readonly router: IRouter;
  readonly logger: Logger;
}

export function defineRoutes(params: RouteDefinitionsParams) {
  defineGetStatsRoute(params);
  defineGetFindingsIndexRoute(params);
```

Comment (kfirpeled) [reply x-pack/plugins/cloud_security_posture/server/routes/index.ts:12]: this way adding a new parameter doesn't require to refactor all signatures

File: x-pack/plugins/cloud_security_posture/server/routes/stats/stats.ts (hunk 4 above)
Comment (kfirpeled) [reply x-pack/plugins/cloud_security_posture/server/routes/stats/stats.ts:52]: no need, just out of curiosity

File: x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts (hunk 6 above)
Comment (kfirpeled) [reply x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts:67]: (sorry for the confusion before)

Comment (CohenIdo) [reply x-pack/plugins/cloud_security_posture/server/index_template/create_index_template.ts:67]: Done:)

File: x-pack/plugins/cloud_security_posture/server/routes/index.ts (hunk 8 above)
Comment (CohenIdo) [reply x-pack/plugins/cloud_security_posture/server/routes/index.ts:12]: You're right, I want to talk with you about this issue.
Is not just the logger, also when we inject other plugins to our plugin and want to enrich `router.context` we can do it with a similar approach (but there is some more approaches in different plugins).
It's required to change the plugin setup and I would like to do it in a different task after merging the benchmark PR as well.

**Comment by CohenIdo** [comment]
The CI is failing due to TS error that was fixed in this [PR](https://github.com/build-security/kibana/pull/102). (Still not merged)

**Review by kfirpeled — DISMISSED** [review]

**Review by kfirpeled — APPROVED** [review]

**Pull request merged**